src/scripts/syncBreaches.js

```diff
@@ -11,7 +11,83 @@
 
 import { req, formatDataClassesArray } from '../utils/hibp.js'
 import { getAllBreaches, upsertBreaches } from '../db/tables/breaches.js'
+import { mkdir, readdir } from "node:fs/promises";
+import { resolve as pathResolve } from "node:path";
+import { finished } from 'stream/promises';
+import fs from "fs";
+import { Readable } from 'stream';
+import aws from 'aws-sdk';
+import { Upload } from "@aws-sdk/lib-storage";
 
+// Get breaches logos and uploads to s3
+const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+const region = process.env.AWS_REGION;
+const Bucket = process.env.S3_BUCKET;
+
+const s3 = new aws.S3({
+  region,
+  credentials: {
+    accessKeyId,
+    secretAccessKey,
+  },
+});
+
+async function uploadToS3(fileName, fileStream) {
+  const uploadParams = {
+    Bucket,
+    Key: fileName,
+    Body: fileStream
+  }
+  try {
+  await s3.upload(uploadParams).promise()
+  console.log('Successfully uploaded data to ' + Bucket + '/' + fileName)
+} catch (err) {
+  console.error(err, err.stack)
+}
+}
+
+export async function getBreachIcons(breaches) {
+  let logoMap;
+  async function fetchBreachIcons() {
+    const breachDomains = breaches
+      .map((breach) => breach.Domain)
+      .filter((breachDomain) => breachDomain.length > 0);
+
+    // make logofolder if it doesn't exist
+    const logoFolder = "public/logo_cache/";
+    if (!fs.existsSync(logoFolder)) await mkdir(logoFolder);
+
+    // read existing logos
+    const existingLogos = await readdir(logoFolder);
+    console.log({existingLogos})
+
+    // TODO: Batch to limit memory use?
+    const logoMapElems = (await Promise.all(
+      breachDomains.map(async (breachDomain) => {
+          const logoFilename = breachDomain.toLowerCase() + ".ico";
+          const logoPath = pathResolve(logoFolder, logoFilename);
+          if (existingLogos.includes(logoFilename)) {
+            console.log('skipping ', logoFilename)
+            return;
+          }
+          console.log(`fetching: , ${logoFilename}`)
+          const res = await fetch(
+            `https://icons.duckduckgo.com/ip3/${breachDomain}.ico`);
+          const fileStream = fs.createWriteStream(logoPath, { flags: 'wx' });
+          const bodyReadable = Readable.fromWeb(res.body)
+          await finished(bodyReadable.pipe(fileStream));
+          await uploadToS3(logoFilename, res.fileStream)
+      })
```

Comment:
Does this need to return a value to the `map()` iterator? I see we have an early exit above on L72, which I assume will return `null` or `undefined` which then gets filtered by the `.filter()` call below on L84. But I'm not clear on what these values would be and how items in `logoMapElems` gets populated.